src/dns_data.py

```diff
@@ -0,0 +1,248 @@
+) -> bool:
+    """Check if a zone definition has changed.
+
+    The zone definition is checked from a custom hash constructed from its DNS records.
+
+    Args:
+        relation_data: input relation data
+        topology: Topology of the current deployment
+
+    Returns:
+        True if a zone has changed, False otherwise.
+    """
+    zones = dns_record_relations_data_to_zones(relation_data)
+    logger.debug("Zones: %s", [z.domain for z in zones])
+    for zone in zones:
+        try:
+            zonefile_content = pathlib.Path(
+                constants.DNS_CONFIG_DIR, f"db.{zone.domain}"
+            ).read_text(encoding="utf-8")
+            metadata = _get_zonefile_metadata(zonefile_content)
+        except (
+            InvalidZoneFileMetadataError,
+            EmptyZoneFileMetadataError,
+            FileNotFoundError,
+        ):
+            return True
+        if "HASH" in metadata and hash(zone) != int(metadata["HASH"]):
+            logger.debug("Config hash has changed !")
+            return True
+
+    if topology is not None and topology.is_current_unit_active:
+        try:
+            named_conf_content = pathlib.Path(
+                constants.DNS_CONFIG_DIR, "named.conf.local"
+            ).read_text(encoding="utf-8")
+        except FileNotFoundError:
+            return True
+        return (
+            _get_secondaries_ip_from_conf(named_conf_content).sort()
+            != [str(ip) for ip in topology.standby_units_ip].sort()
+        )
+
+    return False
+
+
+def record_requirer_data_to_zones(
+    record_requirer_data: DNSRecordRequirerData,
+) -> list[models.Zone]:
+    """Convert DNSRecordRequirerData to zone files.
+
+    Args:
+        record_requirer_data: The input DNSRecordRequirerData
+
+    Returns:
+        A list of zones
+    """
+    zones_entries: dict[str, list[RequirerEntry]] = {}
+    for entry in record_requirer_data.dns_entries:
+        if entry.domain not in zones_entries:
+            zones_entries[entry.domain] = []
+        zones_entries[entry.domain].append(entry)
+
+    zones: list[models.Zone] = []
+    for domain, entries in zones_entries.items():
+        zone = models.Zone(domain=domain, entries=[])
+        for entry in entries:
+            zone.entries.add(models.create_dns_entry_from_requirer_entry(entry))
+        zones.append(zone)
+    return zones
+
+
+def get_conflicts(zones: list[models.Zone]) -> tuple[set[models.DnsEntry], set[models.DnsEntry]]:
+    """Return conflicting and non-conflicting entries.
+
+    Args:
+        zones: list of the zones to check
+
+    Returns:
+        A tuple containing the non-conflicting and conflicting entries
+    """
+    entries = [e for z in zones for e in z.entries]
+    nonconflicting_entries: set[models.DnsEntry] = set()
+    conflicting_entries: set[models.DnsEntry] = set()
+    while entries:
+        entry = entries.pop()
+        found_conflict = False
+        if entry in conflicting_entries:
```

Comment:
I think you can remember this and check if the entry is conflicting, unless there's a big performance penalty, oc

Reply:
After discussion with Arturo, I'm going to revisit this in a later PR